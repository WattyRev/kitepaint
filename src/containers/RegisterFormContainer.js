import React from "react";
import RegisterForm from "../components/RegisterForm";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { CREATE_ACCOUNT } from "../redux/actions";

export class RegisterFormContainer extends React.Component {
  static propTypes = {
    /**
     * A unique identifier to distinguish this from other instances.
     */
    id: PropTypes.string.isRequired,
    /**
     * A function called when the link to go to the log in form is clicked.
     */
    onLogIn: PropTypes.func.isRequired,
    /**
     * A function called when the log in form is submitted.
     * Provided by redux.
     */
    onSubmit: PropTypes.func.isRequired
  };
  state = {
    /**
     * The user entered email address. Used on the registration and reset password forms.
     * @type {String}
     */
    email: "",
    /**
     * Is the register request pending?
     * @type {Boolean}
     */
    pendingRequest: false,
    /**
     * The user entered password. Used on the registration and log in forms.
     * @type {String}
     */
    password: "",
    /**
     * The user entered confirmation password. This is used to force the user to enter the password
     * twice when registering. Used on the registration form.
     * @type {String}
     */
    passwordConfirmation: "",
    /**
     * An error message to display when registration fails.
     * @type {String}
     */
    errorMessage: null,
    /**
     * Indicates that a registration request has been sent. Should be set to true after submission
     * of the registration form, that way we can display a success message to the user.
     * @type {Boolean}
     */
    registrationSent: false,
    /**
     * The user entered username. Used on the registration, log in, and reset password forms.
     * @type {String}
     */
    username: ""
  };
  /**
   * Handles changes to the email field by storing it in state
   * @param  {String} email The new email address
   */
  handleEmailChange = email => {
    this.setState({
      email,
      errorMessage: null
    });
  };

  /**
   * Handles changes to the username field by storing it in state
   * @param  {String} username The new username
   */
  handleUsernameChange = username => {
    this.setState({
      username,
      errorMessage: null
    });
  };

  /**
   * Handles changes to the password field by storing it in state
   * @param  {String} password The new password
   */
  handlePasswordChange = password => {
    this.setState({
      password,
      errorMessage: null
    });
  };

  /**
   * Handles changes to the passwordConfirmation field by storing it in state
   * @param  {String} passwordConfirmation The new passwordConfirmation
   */
  handlePasswordConfirmationChange = passwordConfirmation => {
    this.setState({
      passwordConfirmation,
      errorMessage: null
    });
  };

  /**
   * Handles the log in form submission
   */
  handleSubmit = () => {
    const data = {
      email: this.state.email,
      username: this.state.username,
      password: this.state.password,
      password2: this.state.passwordConfirmation
    };
    this.setState({
      pendingRequest: true
    });
    return this.props
      .onSubmit(data)
      .then(() => {
        this.setState({
          pendingRequest: false,
          registrationSent: true
        });
      })
      .catch(error => {
        this.setState({
          pendingRequest: false,
          errorMessage: error
        });
      });
  };

  render() {
    return (
      <RegisterForm
        email={this.state.email}
        errorMessage={this.state.errorMessage}
        id={this.props.id}
        isDisabled={this.state.pendingRequest}
        onEmailChange={this.handleEmailChange}
        onLogIn={this.props.onLogIn}
        onPasswordChange={this.handlePasswordChange}
        onPasswordConfirmationChange={this.handlePasswordConfirmationChange}
        onSubmit={this.handleSubmit}
        onUsernameChange={this.handleUsernameChange}
        password={this.state.password}
        passwordConfirmation={this.state.passwordConfirmation}
        showSuccessMessage={this.state.registrationSent}
        username={this.state.username}
      />
    );
  }
}

const mapStateToProps = () => ({});

const mapDispatchToProps = {
  onSubmit: CREATE_ACCOUNT
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RegisterFormContainer);
