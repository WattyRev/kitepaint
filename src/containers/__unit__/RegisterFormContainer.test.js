import React from "react";
import { shallow } from "enzyme";
import { RegisterFormContainer } from "../RegisterFormContainer";

describe("RegisterFormContainer", () => {
  let props;
  beforeEach(() => {
    props = {
      id: "abc",
      onLogIn: jest.fn(),
      onSubmit: jest.fn()
    };
  });
  it("renders", () => {
    expect.assertions(1);
    const wrapper = shallow(<RegisterFormContainer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe("handleEmailChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new RegisterFormContainer(props);
      subject.setState = jest.fn();
      subject.handleEmailChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        email: "boogers",
        errorMessage: null
      });
    });
  });
  describe("handleUsernameChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new RegisterFormContainer(props);
      subject.setState = jest.fn();
      subject.handleUsernameChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        username: "boogers",
        errorMessage: null
      });
    });
  });
  describe("handlePasswordChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new RegisterFormContainer(props);
      subject.setState = jest.fn();
      subject.handlePasswordChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        password: "boogers",
        errorMessage: null
      });
    });
  });
  describe("handlePasswordConfirmationChange", () => {
    it("triggers setState with the provided value", () => {
      expect.assertions(2);
      const subject = new RegisterFormContainer(props);
      subject.setState = jest.fn();
      subject.handlePasswordConfirmationChange("boogers");
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        passwordConfirmation: "boogers",
        errorMessage: null
      });
    });
  });
  describe("handleSubmit", () => {
    it("maintains the pendingRequest state value", () => {
      expect.assertions(4);
      props.onSubmit.mockResolvedValue();
      const subject = new RegisterFormContainer(props);
      subject.setState = jest.fn();

      let promise = subject.handleSubmit();
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        pendingRequest: true
      });
      promise = promise.then(() => {
        expect(subject.setState.mock.calls).toHaveLength(2);
        expect(subject.setState.mock.calls[1][0]).toEqual({
          pendingRequest: false,
          registrationSent: true
        });
      });
      return promise;
    });
    it("sets the error message if the call fails", () => {
      expect.assertions(4);
      props.onSubmit.mockRejectedValue("fuck you");
      const subject = new RegisterFormContainer(props);
      subject.setState = jest.fn();

      let promise = subject.handleSubmit();
      expect(subject.setState.mock.calls).toHaveLength(1);
      expect(subject.setState.mock.calls[0][0]).toEqual({
        pendingRequest: true
      });
      promise = promise.then(() => {
        expect(subject.setState.mock.calls).toHaveLength(2);
        expect(subject.setState.mock.calls[1][0]).toEqual({
          pendingRequest: false,
          errorMessage: "fuck you"
        });
      });
      return promise;
    });
  });
});
