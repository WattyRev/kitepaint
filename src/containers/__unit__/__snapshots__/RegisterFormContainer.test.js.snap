// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterFormContainer renders 1`] = `
<RegisterForm
  email=""
  errorMessage={null}
  id="abc"
  isDisabled={false}
  onEmailChange={[Function]}
  onLogIn={[MockFunction]}
  onPasswordChange={[Function]}
  onPasswordConfirmationChange={[Function]}
  onSubmit={[Function]}
  onUsernameChange={[Function]}
  password=""
  passwordConfirmation=""
  showSuccessMessage={false}
  username=""
/>
`;
